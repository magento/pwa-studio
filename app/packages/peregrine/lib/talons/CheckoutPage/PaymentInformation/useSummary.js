import { useQuery } from '@apollo/client';

import { useCartContext } from '../../../context/cart';

import DEFAULT_OPERATIONS from './paymentMethods.gql';
import mergeOperations from '../../../util/shallowMerge';

/**
 * Talon to handle summary component in payment information section of
 * the checkout page.
 *
 * @returns {
 *   isLoading: Boolean,
 *   selectedPaymentMethod: {
 *      code: String,
 *      title: String
 *   }
 * }
 */

export const useSummary = () => {
    const operations = mergeOperations(DEFAULT_OPERATIONS);
    const { getSelectedPaymentMethodQuery } = operations;

    const [{ cartId }] = useCartContext();

    const { data: summaryData, loading: summaryDataLoading } = useQuery(getSelectedPaymentMethodQuery, {
        skip: !cartId,
        variables: { cartId }
    });

    const selectedPaymentMethod = summaryData ? summaryData.cart.selected_payment_method : null;

    return {
        isLoading: summaryDataLoading,
        selectedPaymentMethod
    };
};
