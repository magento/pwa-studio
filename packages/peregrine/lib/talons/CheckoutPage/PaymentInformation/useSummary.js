import { useQuery } from '@apollo/client';

import { useCartContext } from '../../../context/cart';
import mergeOperations from '../../../util/shallowMerge';

import defaultOperations from './summary.gql';
/**
 * Talon to handle summary component in payment information section of
 * the checkout page.
 *
 * @param {DocumentNode} props.queries.getSummaryData gets data from the server for rendering this component
 *
 * @returns {
 *   isLoading: Boolean,
 *   selectedPaymentMethod: {
 *      code: String,
 *      title: String
 *   }
 * }
 */
export const useSummary = (props = {}) => {
    const operations = mergeOperations(defaultOperations, props.operations);

    const { getSummaryData } = operations;

    const [{ cartId }] = useCartContext();

    const { data: summaryData, loading: summaryDataLoading } = useQuery(
        getSummaryData,
        {
            skip: !cartId,
            variables: { cartId }
        }
    );

    const selectedPaymentMethod = summaryData
        ? summaryData.cart.selected_payment_method
        : null;

    return {
        isLoading: summaryDataLoading,
        selectedPaymentMethod
    };
};
