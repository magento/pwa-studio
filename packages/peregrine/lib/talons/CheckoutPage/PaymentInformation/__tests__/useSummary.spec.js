import React from 'react';

import { useQuery } from '@apollo/client';

import { useSummary } from '../useSummary';
import createTestInstance from '../../../../util/createTestInstance';

jest.mock('../../../../context/cart', () => ({
    useCartContext: jest.fn().mockReturnValue([{ cartId: '123' }])
}));

jest.mock('@apollo/client', () => ({
    ...jest.requireActual('@apollo/client'),
    useQuery: jest.fn().mockReturnValueOnce({
        loading: false,
        data: {
            cart: {
                selected_payment_method: 'braintree'
            }
        }
    })
}));

const Component = props => {
    const talonProps = useSummary(props);

    return <i talonProps={talonProps} />;
};

test('Should return correct shape', () => {
    const tree = createTestInstance(<Component queries={{}} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();
});

test('handles empty summary data returned from query', () => {
    useQuery.mockReturnValueOnce({
        loading: false
    });

    const tree = createTestInstance(<Component queries={{}} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps.selectedPaymentMethod).toBeNull();
});
