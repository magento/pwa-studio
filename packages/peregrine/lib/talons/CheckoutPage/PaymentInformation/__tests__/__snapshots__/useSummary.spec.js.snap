// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return correct shape 1`] = `
Object {
  "isLoading": false,
  "selectedPaymentMethod": "braintree",
}
`;
